File: 2872-maximum-number-of-k-divisible-components/2872-maximum-number-of-k-divisible-components.cpp
class Solution {
public:
    void dfs(vector<int>&parent, vector<vector<int>>&child, vector<int>&values, int k, int u, int& cnt) {
        if (child[u].size() != 0) {
            for (int v: child[u]) {
                dfs(parent, child, values, k, v, cnt);
            }
        }
        if (values[u] % k == 0) {
            cnt++;
        }
        else {
            values[parent[u]] = (values[parent[u]] + values[u]) % k;
        }
    }
    int maxKDivisibleComponents(int n, vector<vector<int>>& edges, vector<int>& values, int k) {    
        vector<vector<int>>adj(n, vector<int>());
        for (auto& e: edges) {
            int u = e[0];
            int v = e[1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        vector<int>parent(n, INT_MIN);
        vector<vector<int>>child(n, vector<int>());
        parent[0] = 0;
        queue<int> q;
        q.push(0);
        while (!q.empty()) {
            int u = q.front();
            q.pop();

            for (int v: adj[u]) {
                if (parent[u] != v) {
                    parent[v] = u;
                    child[u].push_back(v);
                    q.push(v);
                }
            }
        }
        int cnt = 0;
        dfs(parent, child, values, k, 0, cnt);
        return cnt;
    }
};

File: 2872-maximum-number-of-k-divisible-components/2872-maximum-number-of-k-divisible-components.py
class Solution:
    def maxKDivisibleComponents(self, n: int, edges: List[List[int]], values: List[int], k: int) -> int:
        adj = [[] for _ in range(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
        
        cnt = 0

        def dfs(parent, u):
            nonlocal cnt
            
            for v in adj[u]:
                if v != parent:
                    dfs(u, v)
            
            if values[u] % k == 0:
                cnt += 1
            else:
                values[parent] = (values[parent] + values[u]) % k
        
        dfs(-1, 0)
        return cnt